### ISSUE TITLE
Include hashtag for notes

### ISSUE THREAD

**Issue by angoca** [Zaczero/openstreetmap-ng#76]
Most of the applications or websites that allows notes creation, append a hashtag providing the application name and in some cases, the version.
Providing this simple information, allow note solver mapper to get a better idea of the note context.

Something like this could be enough

```
#osm-ng
```

**Comment by Zaczero** [comment]
Yup :100:! OpenStreetMap notes should support [hashtags=*](https://wiki.openstreetmap.org/wiki/Key:hashtags) and this is long overdue. Let's start working on it.

The goal is to:
- Add tags dictionary support to notes (just like elements/changesets have it): [this model](https://github.com/Zaczero/openstreetmap-ng/blob/main/app/models/db/note.py) will need to be changed
- Tags are stored in note instance and cannot be modified, they can only be set on note creation
- Have an optional tags input (below the message) on the /note/new page
- Tags input should work like, or use https://github.com/yairEO/tagify - it's very intuitive but maybe there are better options
- Ensure there are no breaking changes in API 0.6 - adding and fetching notes tags will be supported in API 0.7 (which will be dealt with in a separate issue)

**Comment by starsep** [comment]
I agree (hash)tags would be very useful for notes.
I have some UX doubts:
- who will be able to add hashtags? Only note creator when adding note?
- can hashtags be later modified by creator/moderators/other users?

I am thinking about ability to categorize notes.
For example note is valid but requires survey or we need orthoimagery which we expect to be released next year. Or that solving note requires tagging skills in specific area like public transportation.
Perhaps that could be solved by having hashtags in note comments.

As a OSM editor I'd like to filter out notes with some tags. When solving notes I struggled with many notes which are valid (no reason to close them) but difficult to resolve. Some might require extra info etc.

For me the best would be if notes hashtags would work like Github issue tags so they can be modified later when the state of knowledge about note changes after extra comments. But I am not sure about details. It could be also confusing that they would work differently than changeset hashtags

**Comment by Zaczero** [comment]
@starsep Thanks for the feedback! Having read that, I agree it would be beneficial to better manage the state of the note, especially in categorizing notes by other mappers.

What if we store tags data in each NoteComment instance (or None if unchanged) and the current/latest state in the Note instance for efficient database filtering?

When making a comment, you will be able to optionally overwrite the note's tags. By keeping it sequential in NoteComment, we will be able to show modifications over time and quickly revert the state in case of vandalism.

The theoretical API would accept text, action, and tags_or_none. I don't think we would need a versioning system for it since the chance of a conflict would be rare and with low impact. By setting tags dictionary, you completely overwrite the note tags, similarly how uploading a new node/way version works. By not setting one, you don't modify the tags - just a normal text comment.

**Comment by angoca** [comment]
Hashtags in the current OSM form are part of the inmutable text.

If the initial text indicates a particular hashtag (#osmand, #maspme, #hotosm, #colombia, #surveyme), these hashtags can not be modified once created. If, in a comment, another set of hashtags are included, they can not be modified either (#invalid, #solved, #needmoreinfo).

In any case, they are part of the text, not an extra attribute. This is the way it was adapted for API 0.6 to provide some kind of metadata.

Note solvers, like me, we filter notes according hashtags depending on the activity we are doing: auditing, reading discussions, or processing specific projects. 

Having hashtag as extra attributes in the API is a very good proposal idea for API 0.7, and they will be just tags or labels, not #hashtags, and the note consumers should be updated, which will take a lot of time.

For the moment, I just propose to always include a hashtag for the notes created from osm-ng:

```
This the normal text for a note.
#osm-ng
```

Regarding hashtags, there have been some discussions about defining a set of common hashtags for puntual cases. These possible hashtags could be proposed as part of the creation or commenting a note process in osm-ng: https://wiki.openstreetmap.org/wiki/ES:LatAm/Proyectos/Resoluci%C3%B3n_de_notas/hashtags and https://community.openstreetmap.org/t/tagovi-za-odredivanje-statusa-napomena-i-komentara-napomena/4788

**Comment by Zaczero** [comment]
We should parse existing notes hashtags into the new tags format during a migration. We could also include them into notes text when requesting notes via API 0.6 so existing tools remain compatible. I have updated the original work scope.

**Comment by kimbucha** [comment]
I reviewed the expanded structured-note-tag scope in #76 and the existing text-appending PRs. I have a local, AI-assisted extract/append helper with 52 passing isolated tests covering Unicode hashtags, duplicates, URL fragments, Markdown destinations, and stable serialization. It is a foundation for the API/migration work, not a complete implementation of this issue.

Is this scope still wanted, and is the advertised $30 reward still available for accepted contributions? If so, would a focused storage/API increment using the helper be a useful next contribution, and how would an incremental contribution's reward be determined? I can prepare that increment after confirming direction.

**Comment by ikoomm** [comment]
I'll work on a storage/API increment for #76: persist the latest tags on each note and tag-change snapshots on comments, extract hashtags on legacy API writes, and preserve them in legacy responses, search and current UI text. I'll include regression tests for comments that leave tags unchanged and for historical comment output. The dedicated tag editor and backfill of existing notes would remain separate follow-ups.
